### PR TITLE
mgr/dashboard: fix HAProxy (now called ingress)

### PR DESCRIPTION
**Pull request by avanthakkar**
Introducing Ingress service creation from Dashboard UI.

Fixes: https://tracker.ceph.com/issues/50319
Signed-off-by: Avan Thakkar <athakkar@redhat.com>

Ingress Service type creation: 
Screenshot from 2021-04-26 17-55-01

Screenshot from 2021-04-22 15-44-59

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by avanthakkar** [comment]
Dashboard UI now supports creation of Ingress service type which is the current way to have HA endpoint for RGW in cephadm.

**Comment by avanthakkar** [comment]
I've few questions
### Service id: 
As per the docs https://docs.ceph.com/en/latest/cephadm/rgw/#high-availability-service-for-rgw it should follow pattern of existing RGW service, right @sebastian-philipp @jmolmo ? So then should we warn the user to follow the similar pattern as we have for rgw service?
Screenshot from 2021-04-22 15-56-30

### Backend Service:
As currently ingress service allows  to create a high availability endpoint just for RGW (for now) should I expose just RGW services in `Backend Service` field?  I'm showing whole service list.
Screenshot from 2021-04-22 15-35-52

### CIDR (virtual_interface_networks):
It's been exposed as a List[str] in `IngressSpec`, so how it should be represented in UI? We have something similar for 
`Trusted IPs` in `iscsi` service (but it's just a single string, which will not work here, so maybe we can then consume it as `string` too instead of list?)
Screenshot from 2021-04-22 15-59-44

**Comment by avanthakkar** [comment]
jenkins test docs

**Comment by avanthakkar** [comment]
jenkins test dashboard

**Comment by avanthakkar** [comment]
jenkins test dashboard

**Comment by liewegas** [comment]
- The rgw service no longer has the "realm.zone" naming constraint.. it can be named anything.  Instead, there are two optional arguments: `realm` and `zone` that can be passed (together) in a multisite config, or both left off for a single-cluster standalone install
- ``virtual_interface_networks`` is optional.  Normally, if the provided virtual IP is in the same subnet as another IP that is already on the machine, everything Just Works.  However, if the VIP is in a CIDR subnet that doesn't have an existing IP, then the ``virtual_ip_networks`` lets you list other networks that match existing static IPs that are used to identify which interface the new virtual IP should be configured on.

**Comment by tchaikov** [comment]
jenkins test dashboard

**Review by nizamial09 — APPROVED** [review]
Great work @avanthakkar !!

**Comment by avanthakkar** [comment]
Thanks @liewegas for your feedback! Yes it ofc makes sense to prepopulate the `service_id` field and only allow or give a option to customize in case of multiple ingress services required. 

Btw for the implementation of this @epuertat @alfonsomthd @nizamial09  there's a  issue here. As value of any field can't be changed once the form is generated, we need to do it on form creation similar to  [rbd-configuration-form](https://github.com/ceph/ceph/blob/813933f81e3d682a0b1ae6dd906e38e78c4859a4/src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-configuration-form/rbd-configuration-form.component.ts#L112) . We are right now doing almost everything in the html template for `service-form` and I'm not sure how we can prepopulate the value using template (plz do share if u r aware of any such approach). So my proposition here is to let the user manually enter the `Id` field for now(or maybe we can hide it for now?) and then as discussed today in standup, have a follow-up of this feature to include some UI improvements and refactor `service-form` component accordingly.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html [hunk 1]
```diff
@@ -330,8 +360,140 @@
...
+                        [ngValue]="null"
+                        i18n>-- Select an existing RGW service --</option>
+                <option *ngFor="let service of services"
+                        [value]="service.service_name">{{ service.service_name }}</option>
+              </select>
+              <span class="invalid-feedback"
+                    *ngIf="serviceForm.showError('backend_service', frm, 'required')"
+                    i18n>This field is required.</span>
+            </div>
+          </div> -->
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html:392]: Delete?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html [hunk 2]
```diff
@@ -330,8 +360,140 @@
...
+                    i18n>This field is required.</span>
+            </div>
+          </div>
+          <!-- virtual_interface_networks -->
+          <div class="form-group row">
+            <label class="cd-col-form-label"
+                   for="virtual_interface_networks">
+              <span i18n>CIDR Networks</span>
+              <cd-helper>
+                <span i18n>A list of networks to identify which ethernet interface to use for the virtual IP.</span>
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html:484]:
```suggestion
                <span i18n>A list of networks to identify which network interface to use for the virtual IP address.</span>
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html [hunk 3]
```diff
@@ -330,8 +360,140 @@
...
+          </div> -->
+
+          <!-- virtual_ip -->
+          <div class="form-group row">
+            <label class="cd-col-form-label"
+                   [ngClass]="{'required': ['ingress'].includes(serviceForm.controls.service_type.value)}"
+                   for="virtual_ip">
+              <span i18n>Virtual IP</span>
+              <cd-helper>
+                <span i18n>The virtual IP (and network) in CIDR format where the ingress service will be available.</span>
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html:401]:
```suggestion
                <span i18n>The virtual IP address and subnet (in CIDR notation) where the ingress service will be available.</span>
```

**Review by epuertat — APPROVED** [review]
Looks good to me, @avanthakkar ! Great and quick work. Have you considered adding an e2e test?

**Comment by avanthakkar** [comment]
> Looks good to me, @avanthakkar ! Great and quick work. Have you considered adding an e2e test?

Yes, I was planning to add them in a seperate PR as there are none for services at all. So I think maybe all possible e2e tests can be covered in a single PR (including also for service events UI PR by @aaSharma14 ) ?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html (hunk 1 above)
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html:392]: Yes, will do!

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html (hunk 2 above)
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html:484]: Ok, will add this suggestion. I just followed what's mentioned in docs 
https://docs.ceph.com/en/latest/cephadm/rgw/#high-availability-service-for-rgw

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test make check

**Review by alfonsomthd — APPROVED** [review]

**Comment by avanthakkar** [comment]
jenkins test make check

**Pull request merged**